File: 1171-shortest-path-in-binary-matrix/1171-shortest-path-in-binary-matrix.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestPathBinaryMatrix = function(grid) {
    const n = grid.length;
    if(grid[0][0] === 1 || grid[n-1][n-1] === 1){
        return -1;
    }
    
    const dirs = [[1, 0], [0, 1], [-1, 0], [0, -1], [1, 1], [1, -1], [-1, 1], [-1, -1]];
    let dist = 1;

    const queue = [[0, 0]];
    grid[0][0] = 1;
    
    while(queue.length > 0){
        const size = queue.length;

        for(let i=0; i<size; i++){
            const [x, y] = queue.shift();
            if(x === n-1 && y === n-1){
                return dist;
            } 

            for(const [dx, dy] of dirs){
                const nx = x + dx, ny = y + dy;
                if(nx >= 0 && nx < n &&
                ny >= 0 && ny < n &&
                grid[nx][ny] === 0){
                    grid[nx][ny] = 1;
                    queue.push([nx, ny]);
                }
            }
        } 
        dist++;
    }

    return -1;
};

File: 1171-shortest-path-in-binary-matrix/1171-shortest-path-in-binary-matrix.cpp
class Solution {
public:
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        int n = grid.size();
        if(grid[0][0] != 0 || grid[n-1][n-1] != 0) return -1;

        vector<int> dx = {-1,-1,-1, 0,0, 1,1,1};
        vector<int> dy = {-1, 0, 1,-1,1,-1,0,1};
        const int DIRS = 8;

        vector<vector<int>> dist(n, vector<int>(n, 0));

        queue<pair<int, int>> q;
        q.push({0, 0});
        dist[0][0] = 1;

        while(!q.empty()){
            auto [y, x] = q.front(); q.pop();

            if(y == n-1 && x == n-1) return dist[n-1][n-1];

            for(int i=0; i<DIRS; i++){
                int ny = y + dy[i];
                int nx = x + dx[i];

                if(ny < 0 || ny >= n || nx < 0 || nx >= n) continue;

                if(grid[ny][nx] == 0 && dist[ny][nx] == 0){
                    q.push({ny, nx});
                    dist[ny][nx] = dist[y][x] + 1;
                }
            }          
        }

        return dist[n-1][n-1] ? dist[n-1][n-1] : -1;
    }
};